### PR TITLE
feat: added report for mismatch in serial nos and stock quantity in warehouse

### PR DESCRIPTION
**Pull request by AfshanKhan**
Screenshot 2020-07-13 at 5 05 35 PM

### CODE REVIEW

File: erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
+# For license information, please see license.txt
+
+from __future__ import unicode_literals
+import frappe
+from frappe import _
+
+def execute(filters=None):
+	print(filters)
```
Comment (Anurag810) [inline erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.py]: remove print statement

**Review by Anurag810 — CHANGES_REQUESTED** [review]

File: erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.py (hunk 1 above)
Comment (AfshanKhan) [reply erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.py]: Done

File: erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+def get_data(warehouse):
+	serial_item_list = frappe.get_all("Item", filters={
+		'has_serial_no': True,
+	}, fields=['item_code', 'item_name'])
+
+	data = []
+	for item in serial_item_list:
+		total_serial_no = frappe.db.count("Serial No", filters={"item_code": item.item_code, "status": "Active", "warehouse": warehouse})
+
+		if not total_serial_no:
```
Comment (nabinhait) [inline erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.py]: Not required, `frappe.db.count` returns zero if no records

File: erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+			total_serial_no = 0
+
+		actual_qty = frappe.db.get_value('Bin', fieldname=['actual_qty'], filters={"warehouse": warehouse, "item_code": item.item_code})
+
+		if not actual_qty:
```
Comment (nabinhait) [inline erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.py]: Not required, `frappe.db.count` returns zero if no records

File: erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+			"width": 200
+		},
+		{
+			"label": _("Item Name"),
+			"fieldname": "item_name",
+			"fieldtype": "Data",
+			"width": 200
+		},
+		{
+			"label": _("Total no of Serial Nos"),
```
Comment (nabinhait) [inline erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.py]: Serial No Count (smaller label)

File: erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+			"fieldname": "total",
+			"fieldtype": "Float",
+			"width": 150
+		},
+		{
+			"label": _("Balance Qty"),
```
Comment (nabinhait) [inline erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.py]: Stock Qty

File: erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.json [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+ "prepared_report": 0,
+ "ref_doctype": "Item",
+ "report_name": "Mismatch in Serial No count and Item's Stock in Warehouse",
+ "report_type": "Script Report",
+ "roles": [
+  {
+   "role": "Item Manager"
+  },
+  {
+   "role": "Stock Manager"
```
Comment (nabinhait) [inline erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.json]: Keep only Stock User and Stock Manager

File: erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.json [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ "disable_prepared_report": 0,
+ "disabled": 0,
+ "docstatus": 0,
+ "doctype": "Report",
+ "idx": 0,
+ "is_standard": "Yes",
+ "modified": "2020-07-13 16:10:54.030386",
+ "modified_by": "Administrator",
+ "module": "Stock",
+ "name": "Mismatch in Serial No count and Item's Stock in Warehouse",
```
Comment (nabinhait) [inline erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.json]: Rename to "Stock Qty vs Serial No Count"

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+	"formatter": function (value, row, column, data, default_formatter) {
+		value = default_formatter(value, row, column, data);
+		if (column.fieldname == "difference" && data) {
+			if (data.difference > 0) {
+				value = "<span style='color:green'>" + value + "</span>";
+			}
+			else if (data.difference < 0) {
+				value = "<span style='color:red'>" + value + "</span>";
+			}
```
Comment (marination) [inline erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.js]: difference > 0 and difference < 0, both should be red. Green indicates correctness or completion of something, none of which are happening if difference is not 0

File: erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+	return columns
+
+def get_data(warehouse):
+	serial_item_list = frappe.get_all("Item", filters={
+		'has_serial_no': True,
+	}, fields=['item_code', 'item_name'])
+
+	data = []
+	for item in serial_item_list:
+		total_serial_no = frappe.db.count("Serial No", filters={"item_code": item.item_code, "status": "Active", "warehouse": warehouse})
```
Comment (marination) [inline erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.py]: status will include Expired too. We basically have to see if the serial no has a warehouse against it. Active and Expired Serial Nos both belong to a warehouse and contribute to the balance quantity of the Item

File: erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.js (hunk 8 above)
Comment (nabinhait) [reply erpnext/stock/report/mismatch_in_serial_no_count_and_item's_stock_in_warehouse/mismatch_in_serial_no_count_and_item's_stock_in_warehouse.js]: yeah, remove color. does not make any sense

**Review by nabinhait — APPROVED** [review]

**Review by nabinhait — CHANGES_REQUESTED** [review]
Add a link of the report in Stock desk page

**Comment by AfshanKhan** [comment]
> Add a link of the report in Stock desk page

@nabinhait done

File: erpnext/stock/report/stock_qty_vs_serial_no_count/stock_qty_vs_serial_no_count.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+def get_data(warehouse):
+	serial_item_list = frappe.get_all("Item", filters={
+		'has_serial_no': True,
+	}, fields=['item_code', 'item_name'])
+	
+	status_list = ['Active', 'Expired']
+	data = []
+	for item in serial_item_list:
+		total_serial_no = frappe.db.count("Serial No", filters={"item_code": item.item_code, "status": ("in", status_list), "warehouse": warehouse})
```
Comment (rmehta) [inline erpnext/stock/report/stock_qty_vs_serial_no_count/stock_qty_vs_serial_no_count.py]: break into multiple lines

**Review by nabinhait — APPROVED** [review]

**Comment by harshit-30** [comment]
@AfshanKhan please change the name and description of the PR. 

Cc. @nabinhait

**Pull request merged**